### PR TITLE
[core] Improve testing of the adapters

### PR DESCRIPTION
**Pull request by flaviendelangle**
Blocked by https://github.com/mui/material-ui/pull/37073

- 100% code coverage on gregorian adapters (I'll try to reach 100% on jalaali and hijri later)
- Use `toHaveDateTime` whenever possible
- Test the relevant methods with non-iso dates (copied the tests from `@date-io`)
- Cover a few more edge cases in date manipulation
- Sort `testCalculations` in the method order of the adapter

### CODE REVIEW

File: test/utils/addChaiAssertions.ts [hunk 1]
```diff
@@ -7,7 +7,7 @@ declare global {
       /**
        * Matcher with useful error messages if the dates don't match.
        */
-      toEqualDateTime(expected: Date): void;
+      toEqualDateTime(expected: any): void;
```
Comment (flaviendelangle) [inline test/utils/addChaiAssertions.ts:10]: This line was already very wrong, the actual value was `TDate` and now it's `TDate | string`

File: packages/x-date-pickers/src/AdapterMomentJalaali/AdapterMomentJalaali.ts [hunk 2]
```diff
@@ -146,6 +146,10 @@ export class AdapterMomentJalaali extends AdapterMoment implements MuiPickersAda
     return this.moment(value).locale('fa');
   };
 
+  public parseISO = (isoString: string) => {
```
Comment (flaviendelangle) [inline packages/x-date-pickers/src/AdapterMomentJalaali/AdapterMomentJalaali.ts:149]: Was crashing, not sure why
This removed the `strict` attribute but since we are not using this method it's better than having it crash.

File: packages/x-date-pickers/src/AdapterMomentJalaali/AdapterMomentJalaali.ts [hunk 3]
```diff
@@ -186,6 +190,18 @@ export class AdapterMomentJalaali extends AdapterMoment implements MuiPickersAda
     return this.moment(value).isSame(comparing);
   };
 
+  public isSameYear = (value: Moment, comparing: Moment) => {
```
Comment (flaviendelangle) [inline packages/x-date-pickers/src/AdapterMomentJalaali/AdapterMomentJalaali.ts:193]: This was using the `month` and `year` instead of `jMonth` and `jYear` which was just wrong

File: packages/x-date-pickers/src/AdapterDayjs/AdapterDayjs.test.tsx [hunk 4]
```diff
@@ -15,39 +14,25 @@ import {
...
-  },
-  de: {
-    placeholder: 'DD.MM.YYYY hh:mm',
-    value: '15.05.2018 09:35',
-  },
-};
+import 'dayjs/locale/fr';
+import 'dayjs/locale/de';
+// We import it here just to have the typing
+import 'dayjs/plugin/utc';
```
Comment (alexfauquette) [inline packages/x-date-pickers/src/AdapterDayjs/AdapterDayjs.test.tsx:27]: would `import type` change something?

File: packages/x-date-pickers/src/AdapterLuxon/AdapterLuxon.test.tsx [hunk 5]
```diff
@@ -79,8 +63,8 @@ describe('<AdapterLuxon />', () => {
       ) => {
         const date = adapter.date('2020-02-01T23:44:00.000Z')!;
 
-        expect(adapter.format(date, format)).to.equal(expectedWithEn);
-        expect(adapterRu.format(date, format)).to.equal(expectedWithRu);
+        expect(cleanText(adapter.format(date, format))).to.equal(expectedWithEn);
```
Comment (alexfauquette) [inline packages/x-date-pickers/src/AdapterLuxon/AdapterLuxon.test.tsx:66]: Isn't it needed only for russian?

File: packages/x-date-pickers/src/AdapterMoment/AdapterMoment.ts [hunk 6]
```diff
@@ -472,7 +472,7 @@ export class AdapterMoment implements MuiPickersAdapter<Moment, string> {
   public getMeridiemText = (ampm: 'am' | 'pm') => {
     if (this.is12HourCycleInCurrentLocale()) {
       // AM/PM translation only possible in those who have 12 hour cycle in locale.
-      return this.moment
+      return defaultMoment
```
Comment (alexfauquette) [inline packages/x-date-pickers/src/AdapterMoment/AdapterMoment.ts:475]: Why using the default moment instead of the one provided by the user?

**Review by alexfauquette — COMMENTED** [review]
Nice

Is their special stuff to verify in `testCalculations.ts` or is it just reordering and use of `toEqualDateTime`?

File: packages/x-date-pickers/src/AdapterDayjs/AdapterDayjs.test.tsx (hunk 4 above)
Comment (flaviendelangle) [reply packages/x-date-pickers/src/AdapterDayjs/AdapterDayjs.test.tsx:27]: For me no, it would be the same

File: packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts [hunk 7]
```diff
@@ -1,275 +1,529 @@
...
-
-    // Non ISO
-    expect(
-      adapter.formatByString(
-        adapter.startOfWeek(adapter.date('2018-10-28T00:00:00.000Z')!),
-        formatDateTime,
-      ),
-    ).to.equal(adapter.lib === 'luxon' ? '2018-10-22 00:00:00' : '2018-10-28 00:00:00');
+    const expected =
+      adapter.lib === 'luxon' ? '2018-10-29T00:00:00.000Z' : '2018-10-28T00:00:00.000Z';
```
Comment (LukasTy) [inline packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts:359]: Did you consider changing the `adapter` locale in tests to one where weeks start on Monday in order to avoid the need for these checks? 🤔

File: packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts [hunk 8]
```diff
@@ -1,275 +1,529 @@
...
-    expect(adapter.isAfterDay(testDate, nextDay)).to.equal(false);
+  it('Method: getPreviousMonth', () => {
+    expect(adapter.getPreviousMonth(testDateIso)).toEqualDateTime('2018-09-30T11:44:00.000Z');
   });
 
-  it('Method: isBeforeDay', () => {
-    const previousDay = adapter.addDays(testDate, -1);
+  it('Method: getMonthArray', () => {
+    const monthArray = adapter.getMonthArray(testDateIso);
+    let expectedDate = adapter.date('2018-01-01 00:00:00')!;
```
Comment (LukasTy) [inline packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts:498]: Nitpick: it seems that everywhere else the date-time string includes the separating `T` character. 🤔 
```suggestion
    let expectedDate = adapter.date('2018-01-01T00:00:00')!;
```

**Review by LukasTy — APPROVED** [review]
Massive effort. Great job! 🚀 ❤️ 💯

File: packages/x-date-pickers/src/AdapterMoment/AdapterMoment.ts (hunk 6 above)
Comment (flaviendelangle) [reply packages/x-date-pickers/src/AdapterMoment/AdapterMoment.ts:475]: Yes, I'm trying to only use the custom instance to create date, not for the meta data like locales

File: packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts (hunk 7 above)
Comment (flaviendelangle) [reply packages/x-date-pickers/src/tests/describeGregorianAdapter/testCalculations.ts:359]: I like to keep the default locale because it's more broadly known.
But I agree that those exceptions are not great

**Review by alexfauquette — APPROVED** [review]

File: packages/x-date-pickers/src/tests/describeGregorianAdapter/testLocalization.ts [hunk 9]
```diff
@@ -1,14 +1,51 @@
...
     expect(adapter.getMeridiemText('am')).to.equal('AM');
     expect(adapter.getMeridiemText('pm')).to.equal('PM');
+
+    // Moment only translates for 12-hour cycle.
+    if (adapter.lib === 'moment') {
+      const stub = sinon.stub(adapter, 'is12HourCycleInCurrentLocale').returns(false);
+      expect(adapter.getMeridiemText('am')).to.equal('AM');
+      expect(adapter.getMeridiemText('pm')).to.equal('PM');
+      stub.restore();
+    }
```
Comment (LukasTy) [inline packages/x-date-pickers/src/tests/describeGregorianAdapter/testLocalization.ts:25]: Isn't it now the case that only moment adapters are using this test? 🤔 
Shouldn't it have been something like this?
```ts
let stub;
if (adapter.lib === 'moment') {
  stub = sinon.stub(adapter, 'is12HourCycleInCurrentLocale').returns(false);
}
expect(adapter.getMeridiemText('am')).to.equal('AM');
expect(adapter.getMeridiemText('pm')).to.equal('PM');
if (stub) {
  stub.restore();
}
```

Comment (flaviendelangle) [reply packages/x-date-pickers/src/tests/describeGregorianAdapter/testLocalization.ts:25]: No because in order to have 100% test coverage, on moment we want to test both with and without 12 hours cycle.

Comment (LukasTy) [reply packages/x-date-pickers/src/tests/describeGregorianAdapter/testLocalization.ts:25]: Ah, sorry, so, there is another test asserting the logic without a stub..?
In that case, yeah, all is great! 👌 ✅
